src/main/java/seedu/todo/logic/commands/TagCommand.java

```diff
@@ -0,0 +1,238 @@
+        if (isFirstArgNumber) {
+            try {
+                index.setValue(tokens[0]);
+            } catch (IllegalValueException e) {
+                errors.put(index.getName(), e.getMessage());
+            }
+        }
+    }
+
+    @Override
+    public String getCommandName() {
+        return "tag";
+    }
+
+    @Override
+    public List<CommandSummary> getCommandSummary() {
+        String addNewTagsFromTaskArgument = index.getName() + " /a tag1 [, tag2, ...]";
+        String deleteTagsFromTaskArgument = "[" + index.getName() + "] /d tag1 [, tag2, ...]";
+
+        return ImmutableList.of(
+            new CommandSummary("Add tags to a task", getCommandName(), addNewTagsFromTaskArgument),
+            new CommandSummary("Delete tags from tasks", getCommandName(), deleteTagsFromTaskArgument)
+        );
+    }
+
+    @Override
+    protected void validateArguments() {
+        //Check if we have enough input arguments
+        if (!isInputParametersAvailable()) {
+            handleUnavailableInputParameters();
+        }
+
+        //Check arguments for add tags case
+        if (isAddTagsToTask()) {
+            String[] tagsToAdd = StringUtil.splitString(addTags.getValue());
+            checkForIllegalCharInTagNames(addTags.getName(), tagsToAdd);
+            checkForDuplicatedTagNames(addTags.getName(), tagsToAdd);
+        }
+
+        //Check arguments for delete tags case
+        if (isDeleteTagsFromTask()) {
+            String[] tagsToDelete = StringUtil.splitString(deleteTags.getValue());
+            checkForDuplicatedTagNames(deleteTags.getName(), tagsToDelete);
+        }
+        super.validateArguments();
+    }
+
+    @Override
+    public CommandResult execute() throws ValidationException {
+        //Obtain values for manipulation
+        Integer displayedIndex = index.getValue();
+        String[] tagsToAdd = StringUtil.splitString(addTags.getValue());
+        String[] tagsToDelete = StringUtil.splitString(deleteTags.getValue());
+
+        //Performs the actual execution with the data
+        if (isAddTagsToTask()) {
+            model.addTagsToTask(displayedIndex, tagsToAdd);
+            return new CommandResult(StringUtil.convertListToString(tagsToAdd) + SUCCESS_ADD_TAGS);
+
+        } else if (isDeleteTagsFromTask()) {
+            model.deleteTagsFromTask(displayedIndex, tagsToDelete);
+            return new CommandResult(StringUtil.convertListToString(tagsToDelete) + SUCCESS_DELETE_TAGS);
+
+        } else {
+            //Invalid case, should not happen, as we have checked it validateArguments.
+            //However, for completeness, a command result is returned.
+            throw new ValidationException(ERROR_INCOMPLETE_PARAMETERS);
+        }
+    }
+
+    /* Input Parameters Validation */
+    /**
+     * Returns true if the command matches the action of adding tag(s) to a task.
+     */
+    private boolean isAddTagsToTask() {
+        boolean isIndexAvailable = index.getValue() != null;
```

Comment:
You'll probably want `hasBoundValue` instead. Also these variables can probably be inlined